### PR TITLE
[1.1.x] G33 kinematic auto tune factors

### PR DESCRIPTION
**Pull request by LVD-AC**
•updated convergence matrices
•kinematic auto tune scaling parameters
•Z-offset calibration with G33 P-1
•S(et-up) parameter

Counterpart to #8822
[Concise Diff](8821/files?w=1)

### CODE REVIEW

File: Marlin/Marlin.h [hunk 1]
```diff
@@ -387,8 +387,8 @@ void report_current_position();
   extern float z_endstop_adj;
 #endif
 
+extern float zprobe_zoffset;
```
Comment (thinkyhead) [inline Marlin/Marlin.h]: This should not be defined all the time.

Comment (thinkyhead) [inline Marlin/Marlin.h]: This should not be defined all the time.

**Review by thinkyhead — CHANGES_REQUESTED** [review]

File: Marlin/Marlin_main.cpp [hunk 2]
```diff
@@ -3009,12 +3013,12 @@ static void homeaxis(const AxisEnum axis) {
...
     // Delta homing treats the axes as normal linear axes.
 
-    // retrace by the amount specified in delta_endstop_adj + additional 0.1mm in order to have minimum steps
+    // retrace by the amount specified in delta_endstop_adj + additional dist in order to have minimum steps
     if (delta_endstop_adj[axis] * Z_HOME_DIR <= 0) {
       #if ENABLED(DEBUG_LEVELING_FEATURE)
         if (DEBUGGING(LEVELING)) SERIAL_ECHOLNPGM("delta_endstop_adj:");
       #endif
-      do_homing_move(axis, delta_endstop_adj[axis] - 0.1 * Z_HOME_DIR);
+      do_homing_move(axis, delta_endstop_adj[axis] - MIN_STEPS_PER_SEGMENT / planner.axis_steps_per_mm[axis] * Z_HOME_DIR);
```
Comment (thinkyhead) [inline Marlin/Marlin_main.cpp]: Clever maths. I like.

**Review by thinkyhead — APPROVED** [review]

**Comment by thinkyhead** [comment]
Reviewing this now. You've cajoled me into it!

Just kidding. I finally have time to get into this stuff over the next few days.
Let's try and get this to a happy place for all the delta kiddies out there.

File: Marlin/configuration_store.cpp [hunk 3]
```diff
@@ -467,10 +475,9 @@ void MarlinSettings::postprocess() {
 
     _FIELD_TEST(zprobe_zoffset);
 
-    #if !HAS_BED_PROBE
+    #if !HAS_BED_PROBE && !ENABLED(DELTA_AUTO_CALIBRATION)
       const float zprobe_zoffset = 0;
     #endif
-    EEPROM_WRITE(zprobe_zoffset);
```
Comment (thinkyhead) [inline Marlin/configuration_store.cpp]: I'm confused. Why change the condition that creates a dummy `zprobe_zoffset` but then not write it to EEPROM?

File: Marlin/configuration_store.cpp [hunk 4]
```diff
@@ -989,10 +996,9 @@ void MarlinSettings::postprocess() {
 
       _FIELD_TEST(zprobe_zoffset);
 
-      #if !HAS_BED_PROBE
+      #if !HAS_BED_PROBE && !ENABLED(DELTA_AUTO_CALIBRATION)
         float zprobe_zoffset;
       #endif
-      EEPROM_READ(zprobe_zoffset);
```
Comment (thinkyhead) [inline Marlin/configuration_store.cpp]: I'm confused. Why change the condition that creates a dummy `zprobe_zoffset` but then not read it from EEPROM?

What's the bigger picture here? Why should we not save `zprobe_zoffset` to EEPROM anymore for all configurations?

**Comment by LVD-AC** [comment]
"These conflicts are to complex to resolve in the web editor"

I guess this one died of old age

**Comment by thinkyhead** [comment]
> I guess this one died of old age

_The glass if half full…_ I'll resolve them using the tools of the trade. Hold on…

**Comment by thinkyhead** [comment]
Wow, that was surprisingly easy. No conflicts on rebase!
```
git checkout LVD-AC/kinematic-auto-tune-1.1.7 -b bf1_g33_rebase_fun
git rebase -i upstream/bugfix-1.1.x
git push -f LVD-AC HEAD:kinematic-auto-tune-1.1.7
```
[Concise Diff](8821/files?w=1)

**Comment by thinkyhead** [comment]
I've split out the example config changes and extra language changes into separate commits so it's easier to see the important changes by looking at the first commit.
```sh
git diff HEAD HEAD~1 | git apply     ;# UNDO all files to before original commit
                                      # (Github Desktop: Commit undo of just configs and languages)
git reset --hard HEAD                ;# throw away uncommitted changes
git diff HEAD HEAD~1 | git apply     ;# revert files to the original commit's state
                                      # (Github Desktop: Commit example config changes)
                                      # (Github Desktop: Commit language changes)
git rebase -i upstream/bugfix-1.1.x  ;# Let's squash commits!
                                      # (JOE: Combine the undo with the original commit)
git push -f LVD-AC HEAD:kinematic-auto-tune-1.1.7
```

**Comment by thinkyhead** [comment]
`M212` — Agreed! This is more standard. We'll have to retain `M851` for legacy reasons. It will be simple enough. It would just subtract `Z_PROBE_OFFSET_FROM_EXTRUDER` from the given `Z` parameter and then do the same as `M212`.

**Comment by LVD-AC** [comment]
2 G-codes to set the new variable offset:

M212 = M851 - Z_PROBE_OFFSET_FROM_EXTRUDER right?

easy enough...

**Comment by thinkyhead** [comment]
I've been going over this PR tonight and will get back into it more tomorrow. I hope to find the means to integrate this functionality with the least conceptual disruption (paradigm shift), and will aim for getting that done in the next 2-3 days. I do think this is generally cleaner than re-introducing the `home_offset`, but it definitely obfuscates the more familiar "nozzle-to-probe" offset.

I checked the printrbot fork, and `M212` is actually a complete synonym for `M851` (but it also supports X and Y offsets). Although the reprap wiki refers to it as an "offset to home," printrbot specifically refers to the probe that moves with the hotend, and in answers to user questions they refer to it as the "probe-to-nozzle distance" as opposed to a "home offset."

That isn't to say that we can't pick another code or add an extra parameter to accomplish what `M212` is doing in this PR, but it demonstrates that the probe-nozzle relationship is fundamental to much of the thinking around these kinds of adjustments. So we do need to retain functions that specifically relate to adjusting that distance.

Anyway, it's been a long night, so I'll get back into this later.

**Comment by LVD-AC** [comment]
What you say is valid only for people that have a probe. What this PR does it enables babystepping and a G-code command with which people can change the print-height of the 1st layer without disturbing calibration and level data **on all printers**. M851 and M212 are quit synonym indeed. The difference lies IMO that M851 is probe related and a s such you need a probe on your printer to have access to that command, while M212 is general for all printers.

As you will see I only enabled M851 for printers that home with a probe; and enabled M212 for all. I've chosen that option because for legacy reasons that functionality in 1.1.8 only works properly on printers that home with a probe. It used to work as well on deltas with a probe, but that was abandoned so I'm not gonna get into it, water under the bridge. And in my opinion even if you set the offset of the probe, having an additional offset to that independent of the probe might be preferred to some users, and if not... wel just ignore M212 exists.

But if you feel M851 should be enabled on all printers that have a probe, be my guest just change the #if statements (it does exactly the same thing as M212 so there will be no problems).

Just as if you prefer to restrict M212 to printers that do not have a probe; that is just adding a couple of #if's to M212

I leave these decisions up to you; as long as all printer have at least one G-code I'm a happy bunny. (personally I prefer 212, it's an easier number to remember).

**Comment by LVD-AC** [comment]
Another option is off course to give all printers access to M851 (as this it the legacy command in Marlin) and forget about M212 all to getter.

I see it as if all printers do have a probe, some have an automated probe that generates an electrical signal when triggered, but those who do not have that use the nozzle in papertests to determine the height to set up their printer. So nozzle-to-probe offset could be defined to be inclusive to nozzle-to-papertest offset. **M851 for all** is my preferred option, since we all need to probe the bed one way or another; the probing mechanisms we use quit differ, but dealing with them in the exact same way makes it less complicated to explain and understand. Bl-touch, allen key switch, nozzle switch, papertest, whatever... you can babystep or use M851 to tweak the height of your 1st layer. (and it does not mess up previous obtained calibration and bed level data).

**Comment by thinkyhead** [comment]
> I see it as if all printers do have a probe…

Yes! I totally agree, which is how `PROBE_MANUALLY` became a "probe."

**Comment by thinkyhead** [comment]
Rebased this 4 u.

To make this easier to evaluate, I've created a copy where I've reverted all the renaming to "supplemental" with the aim of (1) concealing the usage of "supplemental offset" from the user, (2) incorporating the functionality in a more explicit fashion ("You are now adjusting height and probe offset in conjunction"), and (3) eliminating the treatment of `Z_PROBE_OFFSET_FROM_EXTRUDER` as having any meaningful significance — since users should be free to set this to zero in their configuration and just use `M851` if that is their choice.

From this starting-point I hope to rework the code into a form that accomplishes exactly what this does, but without the seismic shift that this PR imposes. I'll aim to get that done over the next couple of days.

**Comment by LVD-AC** [comment]
Split into #9310 for z-offset and this #8821 for G33 improvements

File: Marlin/Marlin_main.cpp [hunk 5]
```diff
@@ -3106,12 +3106,12 @@ static void homeaxis(const AxisEnum axis) {
...
```
Comment (thinkyhead) [inline Marlin/Marlin_main.cpp:3098]: Note that it's perfectly legal to have `MIN_STEPS_PER_SEGMENT` set to 0.

**Comment by thinkyhead** [comment]
Ok, this seems good! There have been some small changes. I will merge it shortly. Please make a PR with all the same changes for `bugfix-2.0.x` as soon as possible so everything is in sync.

**Comment by LVD-AC** [comment]
Done: see #8822

**Comment by LVD-AC** [comment]
Seems my french side takes over whenever I write centre. LOL

**Comment by thinkyhead** [comment]
Ok, just so I'm clear on this… "Oui" to #10362 and #10363? No special handling for `G33` needed?

**Comment by LVD-AC** [comment]
I see no problems with users setting their own probe depth. G33 S(etup) avoids bowl-shapes so no problems there.

**Pull request merged**